### ISSUE TITLE
Necessity of adding slice sampling code of the parallel Bayesian optimization project

### ISSUE THREAD

**Issue by kejiashi** [scikit-optimize/scikit-optimize#434]
When implementing the parallel Bayesian optimization method, I'm not sure if adding independent slice sampling is needed or not.

The original paper [Snoek et al., 2012] demonstrates the advantage of GP EI MCMC over GP EI per Second. We could also depend on other packages like PyMC3 or only use the function `sample_y` function from `sklearn.gaussian_process` (`gpr.py`).

I've written some code for slice sampling already. Is this necessary?

**Comment by MechCoder** [comment]
To keep a record of our conversation IRL today, we did not use slice sampling initially because we found out just using the maximum log-likelihood estimates gives reasonable performance and is much easier to use. 

I would be interested to see if using slice_sampling provides improvement in performance. In addition, I would be in favour of using existing samplers such as in PyMC3 in order to reduce the burden of code maintenance instead of writing our own implementations but others might have a different opinion. (Sorry that you wrote your own :-|)

**Comment by betatim** [comment]
I am confused. Why compare something like EIps with an approach to sample points in parallel?

My understanding (using EI as a "variable" here):
* if you want to evaluate one point at a time and only care about making progress use EI.
* if you also care about how much it "costs" to make progress use EIps to trade off some "optimality" in terms of EI with cost of getting that evaluation.
* Next is trying to do the whole thing "in parallel", which is where the GP EI MCMC from the paper comes in.

I assume that GP EI MCMC asking for one next point gives you ~the same answer as EI?

The other thing to keep in mind is that we also want to support using trees or NNs as regressors so we should think what methods can be adapted to use regressors that aren't GPs.

**Comment by MechCoder** [comment]
@betatim The difference between GP EI MCMC asking for one point and GP EI is that GP EI uses the maximum log-likelihood estimate of `theta` and `GP EI MCMC` uses `theta` sampled from the `L(theta | data)` using slice sampling (or any other method). See equation 6 of https://papers.nips.cc/paper/4522-practical-bayesian-optimization-of-machine-learning-algorithms.pdf

It would be interesting to see if this approach performs better than just using the maximum log-likelihood estimate since it is a non-convex problem.

**Comment by betatim** [comment]
I'm still confused how EIps comes into this though.

The way I read eqn 7 is that you can plug any acquisition function `a()` into it to make it "parallel".

Same for eqn 6 we can take acquisition function X and make it into "GP X MCMC" by plugging into eqn 6. So if you were using EIps you would make "GP EIps MCMC" by using (6).

**Comment by kejiashi** [comment]
@betatim Excuse me for this late reply. I'm holding the same opinion and wondering if I'm right. (code for [GP EI per second MCMC](https://github.com/JasperSnoek/spearmint/blob/b37a541be1ea035f82c7c82bbd93f5b4320e7d91/spearmint/spearmint/chooser/GPEIperSecChooser.py) and [more choosers](https://github.com/JasperSnoek/spearmint/tree/b37a541be1ea035f82c7c82bbd93f5b4320e7d91/spearmint/spearmint/chooser)) Also there're a few things maybe redundant/noteworthy to attach to this post:

1. Snoek mentions in his DNN method paper that

> Obtaining a closed form expression for the joint acquisi- tion function across multiple inputs is intractable in general (Ginsbourger & Riche, 2010). However, a successful Monte Carlo strategy for parallelizing Bayesian optimization was developed in Snoek et al. (2012). The idea is to marginalize over the possible outcomes of currently running experiments when making a decision about a new experiment to run. Following this strategy, we use the posterior predictive distribution given by Equations 4 and 5 to generate a set of fantasy outcomes for each running experiment which we then use to augment the existing dataset. By averaging over sets of fantasies, we can perform approximate marginalization when computing EI for a candidate point.

2. The reason that I wouldn't consider dependency on `PyMC3` is that the `skopt` package is ~40KB while `PyMC3` package is 20or30MB. And an independent slice sampling algorithm would take up this much space. But I'm not sure about this. 

3. I'm also not sure if other state-of-the-art sampling algorithms like variational inference (ADVI) or No-U-Turn sampler (NUTS) would perform better or be necessary.

**Comment by MechCoder** [comment]
@betatim I think that is right, the MCMC step is independent of the acquisition function. The way I see why they included the EIps calculations in 4a is because to show that even though it includes the compute time, it does not perform really bad.

**Comment by MechCoder** [comment]
@kejiashi 

> I'm also not sure if other state-of-the-art sampling algorithms like variational inference (ADVI) or No-U-Turn sampler (NUTS) would perform better or be necessary.

I agree but I believe I am not qualified enough to know why one sampler would be better than the other.

> The reason that I wouldn't consider dependency on PyMC3 is that the skopt package is ~40KB while PyMC3 package is 20or30MB. 

One can make dependencies optional. It might better to rely on an existing well-tested package to test if the method works and then we can think about writing our own routines. (Yes, open-source is tedious :P )

**Comment by kejiashi** [comment]
@MechCoder Copy that. On realization with PyMC3, will post when finish!

**Comment by betatim** [comment]
Feel free to open a PR once you have "some code", just mark it as "[WIP] title of my PR here". It helps others ask questions as you develop things as well as more time to think about things :)
